assets/js/base/context/providers/cart-checkout/checkout-processor.ts

```diff
@@ -213,97 +219,67 @@ const CheckoutProcessor = () => {
 			  }
 			: {};
 
-		const data = {
-			billing_address: emptyHiddenAddressFields(
-				currentBillingAddress.current
-			),
-			customer_note: orderNotes,
-			create_account: shouldCreateAccount,
-			...paymentData,
-			extensions: { ...extensionData },
-		};
-
-		if ( cartNeedsShipping ) {
-			data.shipping_address = emptyHiddenAddressFields(
-				currentShippingAddress.current
-			);
-		}
-
 		triggerFetch( {
 			path: '/wc/store/v1/checkout',
 			method: 'POST',
-			data,
+			data: {
+				shipping_address: cartNeedsShipping
+					? emptyHiddenAddressFields( currentShippingAddress.current )
+					: undefined,
+				billing_address: emptyHiddenAddressFields(
+					currentBillingAddress.current
+				),
+				customer_note: orderNotes,
+				create_account: shouldCreateAccount,
+				...paymentData,
+				extensions: { ...extensionData },
+			},
 			cache: 'no-store',
 			parse: false,
 		} )
-			.then( ( response ) => {
+			.then( ( response: unknown ) => {
+				assertResponseIsValid( response );
 				processCheckoutResponseHeaders( response.headers );
 				if ( ! response.ok ) {
-					throw new Error( response );
+					throw response;
 				}
-				return response.json();
+				return response.json() as Promise< CheckoutResponseSuccess >;
 			} )
-			.then( ( responseJson ) => {
+			.then( ( responseJson: CheckoutResponseSuccess ) => {
 				__internalProcessCheckoutResponse( responseJson );
 				setIsProcessingOrder( false );
 			} )
-			.catch( ( errorResponse ) => {
+			.catch( ( errorResponse: ApiResponse ) => {
+				processCheckoutResponseHeaders( errorResponse?.headers );
 				try {
-					if ( errorResponse?.headers ) {
-						processCheckoutResponseHeaders( errorResponse.headers );
-					}
 					// This attempts to parse a JSON error response where the status code was 4xx/5xx.
-					errorResponse.json().then( ( response ) => {
-						// If updated cart state was returned, update the store.
-						if ( response.data?.cart ) {
-							receiveCart( response.data.cart );
-						}
-						createErrorNotice(
-							formatStoreApiErrorMessage( response ),
-							{
-								id: 'checkout',
-								context: 'wc/checkout',
-								__unstableHTML: true,
+					errorResponse
+						.json()
+						.then(
+							( response ) => response as CheckoutResponseError
+						)
+						.then( ( response: CheckoutResponseError ) => {
+							if ( response.data?.cart ) {
+								receiveCart( response.data.cart );
 							}
-						);
-						response?.additional_errors?.forEach?.(
-							( additionalError ) => {
-								createErrorNotice( additionalError.message, {
-									id: additionalError.error_code,
-									context: 'wc/checkout',
-									__unstableHTML: true,
-								} );
-							}
-						);
-						__internalProcessCheckoutResponse( response );
-					} );
+							processErrorResponse( response );
```

Comment:
`processErrorResponse` is the new shared util which translates error codes to notices.